src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,103 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int ARRAY_CAPACITY = 10;
+    private T[] elements;
+    private int size = 0;
+
+    public ArrayList() {
+        this.elements = (T[]) new Object[ARRAY_CAPACITY];
+    }
+
+    public ArrayList(T[] elements, int size) {
+        this.elements = elements;
+        this.size = size;
+    }
 
     @Override
     public void add(T value) {
-
+        elements[size] = value;
+        size++;
+        if (elements.length == size) {
+            grow();
+        }
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index >= 0 && index <= size) {
+            if (elements.length == size) {
+                grow();
+            }
+            System.arraycopy(elements, index, elements, index + 1, size - index);
+            elements[index] = value;
+            size++;
+        } else {
+            throw new ArrayIndexOutOfBoundsException();
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        exception(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        exception(index);
+        elements[index] = value;
+        return;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        exception(index);
+        T deletedElement = elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        size--;
+        return deletedElement;
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (t == elements[i]
+                    || ((t != null && elements[i] != null) && t.equals(elements[i]))) {
```

Comment:
```suggestion
                    || t != null && t.equals(elements[i])) {
```
this may be simplified